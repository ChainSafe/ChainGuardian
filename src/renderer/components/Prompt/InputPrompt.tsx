import * as React from "react";
import {InputForm} from "../Input/InputForm";
import {ButtonDestructive, ButtonPrimary} from "../Button/ButtonStandard";
import {useState} from "react";
import {Loading} from "../Loading/Loading";

export interface IInputPromptProps {
    title: string;
    placeholder?: string;
    inputType?: string;
    onSubmit: (data: string) => Promise<ISubmitStatus>;
    onChange?: (e: React.FormEvent<HTMLInputElement>) => void;
    onCancel: () => void;
    display: boolean;
}

export interface ISubmitStatus {
    valid: boolean;
    errorMessage?: string;
}

export const InputPrompt: React.FunctionComponent<IInputPromptProps> = (props: IInputPromptProps) => {
    const [inputData, setInputData] = useState<string>("");
    const [errorMessage, setErrorMessage] = useState<string | undefined>();
    const [valid, setValid] = useState<boolean | undefined>();
    const [loading, setLoading] = useState<boolean>(false);

    async function onSubmitWrapper(): Promise<void> {
        setLoading(true);
        const result = await props.onSubmit(inputData);
        setLoading(false);

        setValid(result.valid);
        if (!result.valid) {
            setErrorMessage(result.errorMessage);
        } else {
            setTimeout(setInputData, 500, ""); /** Prompt reset */
            setTimeout(setValid, 500, undefined);
        }
    }
    function onCancelWrapper(): void {
        setInputData("");
        props.onCancel();
    }

    function handleOnChange(e: React.FormEvent<HTMLInputElement>): void {
        setInputData(e.currentTarget.value);
        if (props.onChange) props.onChange(e);
    }

    return (
        <>
            <div className={`prompt-overlay ${props.display ? "prompt-show" : "prompt-hide"}`}>
                <div className={"prompt-modal"}>
                    <h2 className={"prompt-title"}>{props.title}</h2>
                    <InputForm
                        inputValue={inputData}
                        inputId={"prompt-input"}
                        placeholder={props.placeholder}
                        errorMessage={errorMessage}
                        valid={valid}
                        onChange={handleOnChange}
                        onSubmit={(e): Promise<void> => {
                            e.preventDefault();
                            return onSubmitWrapper();
                        }}
                        type={props.inputType}
                    />
                    <div className={"button-control"}>
                        <div className={"prompt-cancel-button"}>
                            <ButtonDestructive onClick={onCancelWrapper}>Cancel</ButtonDestructive>
                        </div>
                        <div className={"prompt-confirm-button"}>
                            <ButtonPrimary onClick={onSubmitWrapper}>OK</ButtonPrimary>
                        </div>
                    </div>
                </div>
            </div>
            <Loading visible={loading} title='Confirming...' />
        </>
    );
};
