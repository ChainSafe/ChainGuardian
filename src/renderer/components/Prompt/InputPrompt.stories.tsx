import * as React from "react";
import {storiesOf} from "@storybook/react";
import {button, withKnobs} from "@storybook/addon-knobs";
import {InputPrompt, ISubmitStatus} from "./InputPrompt";
import {Background} from "../Background/Background";
import {useState} from "react";

storiesOf("Input prompt", module)
    .add("Basic prompt input", () => {
        const [visible, setVisible] = useState(false);

        const handler = (): void => setVisible(true);
        button("show prompt", handler);

        return (
            <div>
                <InputPrompt
                    display={visible}
                    onSubmit={async (): Promise<ISubmitStatus> => {
                        return {valid: true};
                    }}
                    onCancel={(): void => {
                        setVisible(false);
                    }}
                    title={"Input prompt"}
                />
                <Background />
            </div>
        );
    })
    .addDecorator(withKnobs);
