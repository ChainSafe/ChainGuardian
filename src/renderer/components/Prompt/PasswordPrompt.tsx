import * as React from "react";
import {Keypair} from "@chainsafe/bls";

import {DEFAULT_ACCOUNT} from "../../constants/account";
import database from "../../services/db/api/database";
import {ICGKeystore} from "../../services/keystore";
import {InputPrompt, ISubmitStatus} from "./InputPrompt";

export interface IPasswordPromptProps {
    /**
     * Called on submit password confirmation.
     *
     * @param password - that is entered in password prompt dialog
     */
    onSubmit: (keypair: Keypair) => Promise<void>;
    onChange?: (e: React.FormEvent<HTMLInputElement>) => void;
    /**
     * Called on cancel password prompt.
     */
    onCancel: () => void;
    /**
     * Boolean controlling is prompt displaying.
     */
    display: boolean;
    keystore: ICGKeystore;
}

export const PasswordPrompt: React.FunctionComponent<IPasswordPromptProps> = (props: IPasswordPromptProps) => {
    const onSubmit = async (promptPassword: string): Promise<ISubmitStatus> => {
        const account = await database.account.get(DEFAULT_ACCOUNT);
        if (account != null) {
            const keyPair = await account.unlockKeystore(promptPassword, props.keystore);
            if (keyPair) {
                await props.onSubmit(keyPair);
                return {valid: true};
            } else {
                return {
                    valid: false,
                    errorMessage: "Invalid password",
                };
            }
        } else {
            return {
                valid: false,
                errorMessage: "Error, account not found",
            };
        }
    };

    return (
        <InputPrompt
            onSubmit={onSubmit}
            onCancel={props.onCancel}
            onChange={props.onChange}
            display={props.display}
            title={"Confirm password"}
            placeholder={"Please reenter your password..."}
            inputType={"password"}
        />
    );
};
