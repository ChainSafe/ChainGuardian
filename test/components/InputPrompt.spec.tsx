import * as renderer from "react-test-renderer";
import * as React from "react";
import {InputPrompt, ISubmitStatus} from "../../src/renderer/components/Prompt/InputPrompt";

describe("Input prompt", () => {
    it("renders correctly", () => {
        const tree = renderer
            .create(
                <InputPrompt
                    title={"Make input"}
                    display={true}
                    onSubmit={async (): Promise<ISubmitStatus> => {
                        return {valid: true};
                    }}
                    onCancel={(): void => {}}></InputPrompt>,
            )
            .toJSON();
        expect(tree).toMatchSnapshot();
    });
});
