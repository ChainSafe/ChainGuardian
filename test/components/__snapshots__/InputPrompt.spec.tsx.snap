// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Input prompt renders correctly 1`] = `
Array [
  <div
    className="prompt-overlay prompt-show"
  >
    <div
      className="prompt-modal"
    >
      <h2
        className="prompt-title"
      >
        Make input
      </h2>
      <form
        onSubmit={[Function]}
      >
        <div
          className="label"
        />
        <div
          className="inputform-container"
        >
          <div
            className="input-label"
          />
          <input
            className="inputform  "
            id="prompt-input"
            onChange={[Function]}
            value=""
          />
          <div
            className="input-eye none"
          />
        </div>
        <div
          className="error-message"
        />
      </form>
      <div
        className="button-control"
      >
        <div
          className="prompt-cancel-button"
        >
          <button
            className="button btn-destructive  "
            onClick={[Function]}
          >
            <div
              className=""
            >
              Cancel
            </div>
          </button>
        </div>
        <div
          className="prompt-confirm-button"
        >
          <button
            className="button btn-primary  "
            onClick={[Function]}
          >
            <div
              className=""
            >
              OK
            </div>
          </button>
        </div>
      </div>
    </div>
  </div>,
  <div
    className="loading-container none"
  >
    <div
      className="loading title"
    >
      <h1>
        Confirming...
      </h1>
    </div>
    <div />
    <img
      className="loading-logo"
      src="test-file-stub"
    />
  </div>,
]
`;
